packages/fxa-auth-server/lib/payments/capability.ts

```diff
@@ -229,6 +245,115 @@ export class CapabilityService {
     ];
   }
 
+  /**
+   * Determine the subscription eligibility path for a user for a given plan,
+   * considering existing IAP subscriptions in the process.
+   * Will throw an error if the targetPlanId does not match with a known plan
+   */
+  public async getPlanEligibility(
+    uid: string,
+    targetPlanId: string,
+    useFirestoreProductConfigs: boolean = false
+  ) {
+    const allPlans = await this.stripeHelper.allAbbrevPlans();
+
+    // Create a map of planId: abbrevPlan for speed/ease of lookup later without iterating
+    const allPlansByPlanId: { [key: string]: AbbrevPlan } = allPlans.reduce(
+      (acc, plan) => {
+        return {
+          ...acc,
+          [plan.plan_id]: plan,
+        };
+      },
+      {}
+    );
+
+    const targetPlan = allPlansByPlanId[targetPlanId];
+
+    if (!targetPlan) throw error.unknownSubscriptionPlan(targetPlanId);
+
+    const { productSet: targetProductSet } = productUpgradeFromProductConfig(
+      targetPlan,
+      useFirestoreProductConfigs
+    );
+
+    if (!targetProductSet) {
+      return SubscriptionEligibilityResult.INVALID;
+    }
+
+    // Fetch all user's subscriptions from all sources
+    const stripeSubscriptions = await this.fetchSubscribedPricesFromStripe(uid);
```

Comment:
If it's a quick change, could we put all 3 `await` calls into a `Promise.all`? This shouldn't hold up merging the code though.

Reply:
Added, as well as cleaned up the mapping a little bit. Not too big of a change but still meaningful. Thanks!